resolwe/flow/models/annotations.py

```diff
@@ -0,0 +1,243 @@
+"""Resolwe annotations model."""
+import re
+from enum import Enum
+from typing import TYPE_CHECKING, Any, Dict, List, Optional
+
+from django.core.exceptions import ValidationError
+from django.db import models
+
+from resolwe.permissions.models import PermissionObject
+
+if TYPE_CHECKING:
+    from resolwe.flow.models import Entity
+
+from .base import AuditModel
+
+
+class HandleMissingAnnotations(Enum):
+    """How to handle missing annotations."""
+
+    ADD = "ADD"
+    REMOVE = "REMOVE"
+
+
+class AnnotationType(Enum):
+    """Supported annotation types."""
+
+    DATE = "date"
+    DECIMAL = "decimal"
+    INTEGER = "integer"
+    STRING = "string"
+
+
+class AnnotationGroup(models.Model):
+    """Group of annotation fields."""
+
+    #: the name of the annotation group
+    name = models.CharField(max_length=128)
+
+    #: the label of the annotation group
+    label = models.CharField(max_length=128)
+
+    #: the sorting order among annotation groups
+    sort_order = models.PositiveSmallIntegerField(default=1)
+
+    def __str__(self) -> str:
+        """Return user-friendly string representation."""
+        return f"{self.name}"
+
+    class Meta:
+        """Set the default ordering."""
+
+        ordering = ["sort_order"]
+
+
+class AnnotationField(models.Model):
+    """Annotation field."""
+
+    #: the name of the annotation fields
+    name = models.CharField(max_length=128)
+
+    #: the type of the annotation field
+    type = models.CharField(max_length=16)
+
+    #: the annotation group this field belongs to
+    group = models.ForeignKey(
+        AnnotationGroup, on_delete=models.CASCADE, related_name="fields"
+    )
+
+    #: the sorting order among annotation fields
+    sort_order = models.PositiveSmallIntegerField(default=1)
+
+    #: optional regular expression for validation
+    validator_regex = models.CharField(max_length=128, null=True)
+
+    #: optional map of valid values to labels
+    vocabulary = models.JSONField(null=True)
+
+    #: user visible field name
+    label = models.CharField(max_length=128)
+
+    #: user visible field description
+    description = models.CharField(max_length=256)
+
+    def label_by_value(self, label: str) -> str:
+        """Get the value by label.
+
+        When no value is found the label is returned.
+        """
+        for vocabulary_value, vocabulary_label in self.vocabulary.items():
```

Comment:
My reasoning was that there are usually a very limited number of choices and iteration will be faster than reversing the dictionary  + lookup. I have changed the code to create a reverse dictionary + lookup since it looks nicer.